llvm/include/llvm/ADT/APFloat.h

```diff
@@ -773,13 +791,362 @@ hash_code hash_value(const DoubleAPFloat &Arg);
 DoubleAPFloat scalbn(const DoubleAPFloat &Arg, int Exp, IEEEFloat::roundingMode RM);
 DoubleAPFloat frexp(const DoubleAPFloat &X, int &Exp, IEEEFloat::roundingMode);
 
+class DFPFloat final : public APFloatBase {
+public:
+  /// \name Constructors
+  /// @{
+
+  DFPFloat(const fltSemantics &) {} // Default construct to +0.0
+  DFPFloat(const fltSemantics &, integerPart) { assert(false && "Not Implemented"); }
+  DFPFloat(const fltSemantics &, uninitializedTag) { assert(false && "Not Implemented"); }
+  DFPFloat(const fltSemantics &, const APInt &) { assert(false && "Not Implemented"); }
+  explicit DFPFloat(double d) { assert(false && "Not Implemented"); }
+  explicit DFPFloat(float f) { assert(false && "Not Implemented"); }
+  DFPFloat(const DFPFloat &) { assert(false && "Not Implemented"); }
+  DFPFloat(DFPFloat &&) { assert(false && "Not Implemented"); }
+  ~DFPFloat() {}
+
+  /// @}
+
+  /// Returns whether this instance allocated memory.
+  bool needsCleanup() const { return partCount() > 1; }
+
+  
+  /// \name Arithmetic
+  /// @{
+
+  opStatus add(const DFPFloat &, roundingMode);
+  opStatus subtract(const DFPFloat &, roundingMode);
+  opStatus multiply(const DFPFloat &, roundingMode);
+  opStatus divide(const DFPFloat &, roundingMode);
+  
+  opStatus remainder(const DFPFloat &);
+  
+  opStatus mod(const DFPFloat &);
+  opStatus fusedMultiplyAdd(const DFPFloat &, const DFPFloat &, roundingMode);
+  opStatus roundToIntegral(roundingMode);
+ 
+  opStatus next(bool nextDown);
+
+  /// @}
+
+  /// \name Sign operations.
+  /// @{
+
+  void changeSign();
+
+  /// @}
+
+  /// \name Conversions
+  /// @{
+
+  opStatus convert(const fltSemantics &, roundingMode, bool *);
+  opStatus convertToInteger(MutableArrayRef<integerPart>, unsigned int, bool,
+                            roundingMode, bool *) const;
+  opStatus convertFromAPInt(const APInt &, bool, roundingMode);
+  opStatus convertFromSignExtendedInteger(const integerPart *, unsigned int,
+                                          bool, roundingMode);
+  opStatus convertFromZeroExtendedInteger(const integerPart *, unsigned int,
+                                          bool, roundingMode);
+  Expected<opStatus> convertFromString(StringRef, roundingMode);
+  APInt bitcastToAPInt() const;
+  double convertToDouble() const;
+  float convertToFloat() const;
+
+  /// @}
+
+  
+  bool operator==(const IEEEFloat &) const = delete;
+
+  
+  cmpResult compare(const IEEEFloat &) const;
+
+  
+  bool bitwiseIsEqual(const IEEEFloat &) const;
+
+  
+  unsigned int convertToHexString(char *dst, unsigned int hexDigits,
+                                  bool upperCase, roundingMode) const;
+
+ 
+  bool isNegative() const { return sign; }
+
+  
+  bool isNormal() const { return !isDenormal() && isFiniteNonZero(); }
+
+ 
+  bool isFinite() const { return !isNaN() && !isInfinity(); }
+
+  
+  bool isZero() const { return category == fcZero; }
+
+  
+  bool isDenormal() const;
```

Comment:
DFP types don't support the notion of a subnormal or denormal value. It looks like a function definition will either be needed for dispatch via `APFLOAT_DISPATCH_ON_SEMANTICS` in `APFloat::isDenormal()`, or the latter function will need to be special cased for `IEEEFloat` and `DoubleAPFloat`. Either approach seems reasonable to me. If we do provide a definition, then we need to decide whether it should just return false or should call `llvm::report_fatal_error()` or similar.